From 4487c704c2a01c6ede04c5217f5e3d12c4249935 Mon Sep 17 00:00:00 2001
From: denglei <denglei@eswincomputing.com>
Date: Mon, 18 Aug 2025 17:04:03 +0800
Subject: [PATCH 461/533] fix:Resolve the issue of NPU frame statistics.

Changelogs:
	1.Resolve the issue of NPU frame statistics.
	2.Resolve NPU compilation warning issue.

Signed-off-by: denglei <denglei@eswincomputing.com>
---
 drivers/soc/eswin/ai_driver/npu/npu_frame.c   |  2 --
 drivers/soc/eswin/ai_driver/npu/nvdla_hw.c    | 29 ++++++++++++++++++-
 drivers/soc/eswin/ai_driver/npu/scheduler.c   |  3 +-
 .../soc/eswin/ai_driver/npu/user_context.c    |  1 -
 4 files changed, 29 insertions(+), 6 deletions(-)

diff --git a/drivers/soc/eswin/ai_driver/npu/npu_frame.c b/drivers/soc/eswin/ai_driver/npu/npu_frame.c
index 2a526f5eb81f..ddf2513e5e16 100644
--- a/drivers/soc/eswin/ai_driver/npu/npu_frame.c
+++ b/drivers/soc/eswin/ai_driver/npu/npu_frame.c
@@ -344,10 +344,8 @@ void npu_drop_all_frame(struct nvdla_device *ndev, bool dump)
 static void npu_process_timeout(struct win_engine *engine, u32 tiktok)
 {
 	struct host_frame_desc *f;
-	struct user_model *model;
 	unsigned long flags;
 	struct win_executor *executor;
-	unsigned long last_state;
 
 	spin_lock_irqsave(&engine->executor_lock, flags);
 	engine->engine_is_alive = false;
diff --git a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
index b6d8889bd2e3..1a11a498258d 100644
--- a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
+++ b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
@@ -526,6 +526,12 @@ static ssize_t store_reset_hand(struct device *d, struct device_attribute *attr,
 	mutex_lock(&engine->reset_mutex);
 	if (ptr != NULL) {
 		if (is_reset == 1) {
+			ret = npu_pm_get(nvdla_dev);
+			if (ret < 0) {
+				dla_error("%s, %d, pm get sync err, ret=%d.\n", __func__,
+					__LINE__, ret);
+				return ret;
+			}
 			engine->engine_is_alive = false;
 			npu_drop_all_frame(nvdla_dev, false);
 			memset(engine->host_node, 0, sizeof(host_node_t));
@@ -604,6 +610,7 @@ static ssize_t store_reset_hand(struct device *d, struct device_attribute *attr,
 				goto exit;
 			}
 			engine->engine_is_alive = true;
+			npu_pm_put(nvdla_dev);
 		} else {
 			dla_error("err:input err.\n");
 		}
@@ -612,6 +619,7 @@ static ssize_t store_reset_hand(struct device *d, struct device_attribute *attr,
 exit:
 	npu_clk_reset_print(nvdla_dev->pdev, nvdla_dev->numa_id);
 	mutex_unlock(&engine->reset_mutex);
+	npu_pm_put(nvdla_dev);
 	return count;
 }
 
@@ -619,11 +627,21 @@ static ssize_t npu_drop_frame(struct device *d, struct device_attribute *attr,
 			      const char *buf, size_t count)
 {
 	struct nvdla_device *nvdla_dev = dev_get_drvdata(d);
+	int ret;
 
 	if (!nvdla_dev) {
 		return -ENODEV;
 	}
+
+	ret = npu_pm_get(nvdla_dev);
+	if (ret < 0) {
+		dla_error("%s, %d, pm get sync err, ret=%d.\n", __func__,
+			__LINE__, ret);
+		return ret;
+	}
 	npu_drop_all_frame(nvdla_dev, true);
+	npu_pm_put(nvdla_dev);
+
 	return count;
 }
 
@@ -632,7 +650,7 @@ static ssize_t npu_dump_dtim(struct device *d, struct device_attribute *attr,
 {
 	struct nvdla_device *nvdla_dev = dev_get_drvdata(d);
 	struct win_engine *engine;
-	int i;
+	int ret, i;
 
 	if (!nvdla_dev) {
 		return -ENODEV;
@@ -642,6 +660,13 @@ static ssize_t npu_dump_dtim(struct device *d, struct device_attribute *attr,
 		return -ENODEV;
 	}
 
+	ret = npu_pm_get(nvdla_dev);
+	if (ret < 0) {
+		dla_error("%s, %d, pm get sync err, ret=%d.\n", __func__,
+			__LINE__, ret);
+		return ret;
+	}
+
 	if (engine->master_mem) {
 		memcpy(engine->master_mem, engine->master_shm,
 		       E31_EMISSION_DTIM_SIZE);
@@ -657,6 +682,8 @@ static ssize_t npu_dump_dtim(struct device *d, struct device_attribute *attr,
 		}
 	}
 	dump_dtim_to_file(engine, 3);
+	npu_pm_put(nvdla_dev);
+
 	return count;
 }
 
diff --git a/drivers/soc/eswin/ai_driver/npu/scheduler.c b/drivers/soc/eswin/ai_driver/npu/scheduler.c
index 6e14b3d978cc..436b20387ee1 100644
--- a/drivers/soc/eswin/ai_driver/npu/scheduler.c
+++ b/drivers/soc/eswin/ai_driver/npu/scheduler.c
@@ -62,7 +62,7 @@ static int send_frame_to_hw(struct win_engine *engine, u8 tiktok,
 			    npu_io_tensor_t *io_tensor, host_node_t *host_node)
 {
 	int ret;
-	struct nvdla_device *ndev = (struct nvdla_device *)engine->nvdla_dev;
+
 	emission_node_t *pemission_node =
 		(emission_node_t *)host_node->emission_base_addr;
 	program_node_t *program_node =
@@ -274,7 +274,6 @@ void mbx_irq_frame_done(struct win_engine *priv, u32 tiktok, u32 stat)
 	struct host_frame_desc *f;
 	struct user_model *model;
 	int ret = 1;
-	unsigned long last_state;
 	unsigned long flags;
 	struct nvdla_device *ndev = (struct nvdla_device *)engine->nvdla_dev;
 
diff --git a/drivers/soc/eswin/ai_driver/npu/user_context.c b/drivers/soc/eswin/ai_driver/npu/user_context.c
index 15f59c4a400d..dc4a2a6de0c8 100644
--- a/drivers/soc/eswin/ai_driver/npu/user_context.c
+++ b/drivers/soc/eswin/ai_driver/npu/user_context.c
@@ -529,7 +529,6 @@ static int commit_new_io_tensor(struct user_context *uctx, void *arg)
 	struct win_executor *executor;
 	int ret;
 	bool result;
-	int new_state;
 
 	model = npu_get_model_by_id(uctx, idx);
 	if (model == NULL) {
-- 
2.51.1

